### PR TITLE
chore: Pycasbin benchmark

### PR DESCRIPTION
**Pull request by sheny1xuan**
+ Add benchmark for pycabin binding #181.
+ Pycasbin binding result in my local machine, and the stddev is greater than mean, it's invalid.
image

> pybind11 heavily relies on a template matching mechanism to convert parameters and return values that are constructed from STL data types such as vectors, linked lists, hash tables, etc. This even works in a recursive manner, for instance to deal with lists of hash maps of pairs of elementary and custom types, etc.
However, a fundamental limitation of this approach is that internal conversions between Python and C++ types involve a copy operation that prevents pass-by-reference semantics. 
+ Current pybind11 binding copy python list to a vector, it's will be inefficient. And I think wrap `DataVector` and `DataList` with python will be helpful. We can ref to https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html.
+ And another thing, current casbin-cpp efficiency is smiliar with casbin-golang. Can we make it better? I think I need some help for the deep optimization. @EmperorYP7 @hsluoyz

### CODE REVIEW

**Comment by hsluoyz** [comment]
@EmperorYP7 plz review

File: pycasbin/README.md [hunk 1]
```diff
@@ -82,6 +82,12 @@ def isAuthorized(req):
 
 Rest of the method's name is on par with `casbin-cpp`.
 
+### Benchmark
+
+Pycasbin use `pytest` for benchmark.
```
Comment (EmperorYP7) [inline pycasbin/README.md:87]:
```suggestion
Pycasbin use `pytest` for benchmark.

`pip3 install -U pytest`
```

File: pycasbin/README.md [hunk 2]
```diff
@@ -82,6 +82,12 @@ def isAuthorized(req):
...
+
+Install `pytest` and `pycasbin` in your local machine, then run the benchmark by `python3 -m pytest --benchmark-verbose --benchmark-columns=mean,stddev,iqr,ops,rounds casbin-cpp/pycasbin/benchmarks/benchmark_model.py`.
```
Comment (EmperorYP7) [inline pycasbin/README.md:89]: Do this instead:

```bash
python3 -m pytest --benchmark-verbose --benchmark-columns=mean,stddev,iqr,ops,rounds casbin-cpp/pycasbin/benchmarks/benchmark_model.py
```

File: pycasbin/benchmarks/benchmark_model.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    e.add_grouping_policies(
+        {("user" + str(i), "group" + str(int(i / 10))) for i in range(10000)}
+    )
+
+    @benchmark
+    def benchmark_rbac_model():
+        e.enforce("user501", "data9", "read")
+
+
+# TODO: pycasbin cost too much time in large model
```
Comment (EmperorYP7) [inline pycasbin/benchmarks/benchmark_model.py:91]: Does it crash often?

How about we set up a flag through an OS env variable for intensive testing?

**Review by EmperorYP7 — CHANGES_REQUESTED** [review]

**Comment by hsluoyz** [comment]
@sheny1xuan

**Comment by ArashPartow** [comment]
@sheny1xuan  in the first set of timing numbers, the stddev is greater than the mean. 

This usually indicates invalid stats or the possibility of at least a bimodal (or more) distribution.

**Comment by hsluoyz** [comment]
@sheny1xuan

**Comment by sheny1xuan** [comment]
> @sheny1xuan in the first set of timing numbers, the stddev is greater than the mean.
> 
> This usually indicates invalid stats or the possibility of at least a bimodal (or more) distribution.

Oh, I only paid attention to the mean, the stddev is also very important in the benchmark. And the above comparison of pycasbin and casbin-cpp binding is valid. The two set of benchmark data is all about casbin-cpp binding, I forget to change the package install environment.

File: pycasbin/benchmarks/benchmark_model.py (hunk 3 above)
Comment (sheny1xuan) [reply pycasbin/benchmarks/benchmark_model.py:91]: It cost too much time in my machine, don't crash.

**Pull request closed without merging**